src/neo4j_graphrag/llm/openai_llm.py

```diff
@@ -121,6 +125,79 @@ def invoke(
         except self.openai.OpenAIError as e:
             raise LLMGenerationError(e)
 
+    def invoke_with_tools(
+        self,
+        input: str,
+        tools: List[dict[str, Any]],  # Tools definition as a list of dictionaries
+        message_history: Optional[Union[List[LLMMessage], MessageHistory]] = None,
+        system_instruction: Optional[str] = None,
+    ) -> ToolCallResponse:
+        """Sends a text input to the OpenAI chat completion model with tool definitions
+        and retrieves a tool call response.
+
+        Args:
+            input (str): Text sent to the LLM.
+            tools (List[Dict[str, Any]]): List of tool definitions for the LLM to choose from.
+            message_history (Optional[Union[List[LLMMessage], MessageHistory]]): A collection previous messages,
+                with each message having a specific role assigned.
+            system_instruction (Optional[str]): An option to override the llm system message for this invocation.
+
+        Returns:
+            ToolCallResponse: The response from the LLM containing a tool call.
+
+        Raises:
+            LLMGenerationError: If anything goes wrong.
+        """
+        try:
+            if isinstance(message_history, MessageHistory):
+                message_history = message_history.messages
+
+            params = self.model_params.copy() if self.model_params else {}
+            if "temperature" not in params:
+                params["temperature"] = 0.0
+
+            # Convert tools to OpenAI's expected type
+            openai_tools: List[ChatCompletionToolParam] = []
+            for tool in tools:
+                openai_tools.append(cast(ChatCompletionToolParam, tool))
```

Comment:
Can't we create the object with `ChatCompletionToolParam(**tool)` (or something similar)?

Reply:
Yes I guess this is because `openai_tools` is declared to be `List[ChatCompletionToolParam]` but `tool` is a dict. That's why I thought instantiating the `ChatCompletionToolParam` could fix mypy without having to use `cast`, but I don't know the model exactly so maybe it doesn't work.